accepted/00062-hub-xmlrpc-api.md

```diff
@@ -0,0 +1,136 @@
+
+* Authorization and authentication
+  * Gateway API will be secured via `hubSessionKey` tokens analoguously to Server's API
+    * Actually, Gateway will delegate authentication to the Hub API
+  * Consuming a Server's API from the Gateway continues to require a valid Server `sessionKey` token
+  * There are three ways to attach Server sessions to the Hub session
+    * Manual mode: programmer "attaches" Server sessions explicitly to his Hub session. Attachment lasts until logout
+      * `hub.login(username, password)` → `hubSessionKey`
+      * `hub.attachToServer(hubSessionKey, serverId, username, password)` → `serverSessionKey`
+      * `hub.serverMethod(serverSessionKey, parameters)` → `output`
+      * N+1 pairs of credentials will be needed (one for the Hub, one per each of the N Servers)
+     * Authentication relay mode: as an additional convenience option, which can be activated with a flag, Hub-User credentials can be re-used to authenticate against Servers. So if a certain username/password is uniformly configured on the Hub and across several Servers, Gateway usage becomes simpler: only one log in is needed, with its Hub-User credentials
+     * Automatic connect mode: as an additional convenience option, which can be activated with a flag, the Hub gateway API can automatically select the list of Servers a Hub-User works with, based on the list of Servers that same Hub-User has permissions on
+       - note: that this only makes sense if users are OK with the idea of using Server sysadmin users to act on Clients
+       - If they prefer to manage Clients via users that have no permissions to manage Servers, they can create a Hub-Org with no Servers. Any Hub-User in this Hub-Org will be able to connect to the Gateway, and possibly have permissions on Clients (if Servers have Server-Users with same name and passwords)
+
+* Multicasting: call an XMLRPC method on _N_ Servers at once
+  * method name stays the same
+  * Assuming the method on the Server accepts _M_ parameters, the same method on the Hub will accept an array of _N_ parameters, each of which contains as elements the _M_ parameters for a given Server
+    * eg. `actionchain.listChains(sessionKey)` → `hub.actionchain.listChains([serverSessionKey1, serverSessionKey2, ...])`
+    * eg. `channel.org.enableAccess(sessionKey, channelLabel, orgId)` → `hub.channel.org.enableAccess([serverSessionKey1, channelLabel1, orgId1], [serverSessionKey2, channelLabel2, orgId2], ...)`
+  * return types would be arrays of _N_ elements
+    * special elements would be needed to signal Exceptions
+
+* Availability
+  * if a Server is down at the time a Gateway call targeting it is made, call should fail after a configurable timeout. If multiple Servers are targeted, only that call fails and others continue
```

Comment:
IMO this should be managed differently: Server API calls should be classified in several categories according to the required response time and the hub API should do the retries.

Use case: it's 10 AM in the morning and I am scheduling an update cycle for tonight at 11 PM. I have 10 Uyuni Servers and 1 of them is down at 10:00:56. Instead of returning failure, Hub should keep retrying until e. g. 10 PM => different API calls have different timeout needs.

E. g. if I want to run a command on all my clients, I want the response to be immediate (i. e. not more than a very short timeout, maybe 1 minute) => different API calls have different timeout needs

Reply:
One thing that could be done is to make it configurable per-call. I will be amending the text in this sense in a follow-up RFC, thanks for pointing out.